sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala

```diff
@@ -180,13 +180,15 @@ case class CaseWhen(
   }
 
   override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
-    // This variable represents whether the first successful condition is met or not.
-    // It is initialized to `false` and it is set to `true` when the first condition which
-    // evaluates to `true` is met and therefore is not needed to go on anymore on the computation
-    // of the following conditions.
-    val conditionMet = ctx.freshName("caseWhenConditionMet")
-    ctx.addMutableState(ctx.JAVA_BOOLEAN, ev.isNull)
-    ctx.addMutableState(ctx.javaType(dataType), ev.value)
+    // This variable represents whether the evaluated result is null or not. It's a byte value
+    // instead of boolean because it carries an extra information about if the case-when condition
+    // is met or not. It is initialized to `-1`, which means the condition is not met yet and the
+    // result is unknown. When the first condition is met, it is set to `1` if result is null, or
+    // `0` if result is not null. We won't go on anymore on the computation if it's set to `1` or
+    // `0`.
+    val resultIsNull = ctx.freshName("caseWhenResultIsNull")
```

Comment:
I think there is no need to change the type of this. We can achieve the reduction of number of variables used as you did for coalesce, thus maintaining consistency with the rest of the code

Reply:
Could you elaborate your thought on how to use `boolean` in the case `CaseWhen`? 

IIUC,  `coalesce` requires two states: `not met with isNull = true` or `met with isNull = false`. Thus, we can use `boolean` value. `CaseWhen` requires three states: `not met`, `met with isNull = false`, or `met with isNull = true`. Thus, we introduces `byte` value. If we would use `boolean value` like `coalesce`, we will miss the state `met with isNull = true`.

Reply:
I see, you are right, sorry. The names were misleading me. Can we change the name then? In the comment, can we have a schematic description like:
```
This variable holds the state of the operation:
 - -1 means ...
 - 0 means ...
 - 1 means ...
```
Thanks.

Reply:
thanks, this is based on [review comments](https://github.com/apache/spark/pull/19901#discussion_r155131706).

Reply:
I see, thanks for the answer. Because the point is that this variable doesn't carry only the information about being null or not. If @cloud-fan confirms that these are good names, it is ok, but I'd prefer something more explicit.

Reply:
I agree `resultIsNull` is not a perfect name, but I don't have better idea, maybe `resultIsNullOrConditionMet`?

Reply:
or `resultState`?